flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,308 @@
+        // we track the job termination via all execution termination futures (FLINK-21030).
+        final CompletableFuture<Collection<ExecutionState>> executionTerminationsFuture =
+                SchedulerUtils.getCombinedExecutionTerminationFuture(executionGraph);
+
+        // do not trigger checkpoints while creating the final savepoint
+        stopCheckpointScheduler();
+
+        // trigger savepoint. This operation will also terminate/suspend the job once the savepoint
+        // has been created.
+        final CompletableFuture<CompletedCheckpoint> savepointFuture =
+                triggerSynchronousSavepoint(terminate, targetDirectory);
+
+        final StopWithSavepointTerminationManager stopWithSavepointTerminationManager =
+                new StopWithSavepointTerminationManager(
+                        new StopWithSavepointTerminationHandlerImpl(
+                                executionGraph.getJobID(), context, this, logger));
+
+        this.operationCompletionFuture =
+                stopWithSavepointTerminationManager.trackStopWithSavepoint(
+                        savepointFuture,
+                        executionTerminationsFuture,
+                        context.getMainThreadExecutor());
+    }
+
+    @Override
+    public void cancel() {
+        // the canceling state will cancel the execution graph, which will eventually lead to a call
+        // of handleAnyFailure(), completing the operationCompletionFuture.
+        context.goToCanceling(
+                getExecutionGraph(), getExecutionGraphHandler(), getOperatorCoordinatorHandler());
+    }
+
+    @Override
+    public JobStatus getJobStatus() {
+        return JobStatus.RUNNING;
+    }
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        handleAnyFailure(cause);
+    }
+
+    @Override
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
+        final boolean successfulUpdate =
+                getExecutionGraph().updateState(taskExecutionStateTransition);
+
+        if (successfulUpdate) {
+            if (taskExecutionStateTransition.getExecutionState() == ExecutionState.FAILED) {
+                Throwable cause = taskExecutionStateTransition.getError(userCodeClassLoader);
+                handleAnyFailure(cause);
+            }
+        }
+
+        return successfulUpdate;
+    }
+
+    @Override
+    void onGloballyTerminalState(JobStatus globallyTerminalState) {
+        context.goToFinished(ArchivedExecutionGraph.createFrom(getExecutionGraph()));
+    }
+
+    private void handleAnyFailure(Throwable cause) {
+        operationCompletionFuture.completeExceptionally(
+                new FlinkException("Failure while stopping with savepoint", cause));
+        final Executing.FailureResult failureResult = context.howToHandleFailure(cause);
+
+        if (failureResult.canRestart()) {
+            startCheckpointScheduler();
```

Comment:
We don't have to explicitly enable Checkpoint scheduling here. This is done by the `CheckpointCoordinatorDeActivator` after the `JobStatus` changes from `RESTARTING` to `RUNNING` again.